### DISCUSSION TITLE
example integrations with other routers

### DISCUSSION THREAD

**Discussion by c-nv-s** [doytowin/goooqo#15] [General]
could you please show some examples of integrating goooqo with other routers e.g. echo, chi

what if a developer wants to place middleware in front of the POST/PUT/PATCH methods, for example to check for an Authorization header... please show an example of that.

**Comment by f0rb** [comment]
OK, I will make an integrating example with Echo later in the demo repo.

There are too many good web frameworks in Go to choose from, so I made a simple one to evaluate my thoughts.  
For placing middleware, you can refer to your integrating routers I guess. 
GoooQo will just focus on data access layer.

**Comment by f0rb** [comment]
I have pushed the code for Echo: https://github.com/doytowin/goooqo-demo/tree/echo
```
	e := echo.New()
	buildEchoService[UserEntity, UserQuery](e, "/user/", userDataAccess)
	e.Logger.Fatal(e.Start(":9090"))
```
Almost the same with the original demo.
I just copied the `func resolveQuery(queryMap url.Values, query any)` instead of using `(&DefaultBinder{}).BindQueryParams(c, &payload)` since it needs extra tags on the struct. And I think it doesn't support nested struct and path well.

**Comment by c-nv-s** [comment]
Thanks a lot for coming back with that.
looks like the library's API is a little more involved for integration with other routers but still is very handy.
I'll be seeing if I can use it with a few of my projects.

I think the icing on the cake now is some form of https://github.com/ogen-go/ogen integration to auto-generate the API docs

**Comment by f0rb** [comment]
Yes， I think the complexity of integration comes from two aspects：
1. Different routers provide their own APIs to bind HTTP methods, paths, and handlers, so I think there should be a module to bind the RESTful APIs for each router.
2. Not all routers provide an out-of-box method to map query parameters to the query struct, so I make the ResolveQuery method public as a workaround.

And thanks for your information, I will check how Ogen works. 
Maybe I can provide a module to integrate goooqo with Ogen and different routers. Not sure yet.
